### PR TITLE
[UKT0] Add support for Keystone authentication via EGI's Dev AAI

### PR DESCRIPTION
**Pull request by yankcrime**
Adds support for user authentication via the **development** instance of [EGI's Check-in AAI](https://wiki.egi.eu/wiki/AAI) service.

This is part of the work being done for UKT0 and the federation of OpenStack resources, for which StackHPC is developing a POC deployment that will include AlaSKA.

In the first instance these changes will be confined to the `alt-1` partition.

### CODE REVIEW

**Review by JohnGarbutt — APPROVED** [review]
This seems OK to me, but its tricky to review given the tweaks on top of the "default", but not seeing the diff. Must find a better way to package that longer term.

File: etc/kayobe/kolla/config/keystone/wsgi-keystone.conf [hunk 1]
```diff
@@ -0,0 +1,63 @@
+{% raw %}
+{% set keystone_log_dir = '/var/log/kolla/keystone' %}
+{% set python_path = '/usr/lib/python2.7/site-packages' if kolla_install_type == 'binary' else '/var/lib/kolla/venv/lib/python2.7/site-packages' %}
+Listen {{ api_interface_address }}:{{ keystone_public_port }}
+Listen {{ api_interface_address }}:{{ keystone_admin_port }}
+
+TraceEnable off
```
Comment (dougszumski) [inline etc/kayobe/kolla/config/keystone/wsgi-keystone.conf:7]: Perhaps worth having a comment separating the change you've made from the base wsgi conf file?

File: etc/kayobe/kolla/config/keystone/wsgi-keystone.conf [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    LogFormat "%{X-Forwarded-For}i %l %u %t \"%r\" %>s %b %D \"%{Referer}i\" \"%{User-Agent}i\"" logformat
+    CustomLog "{{ keystone_log_dir }}/keystone-apache-public-access.log" logformat
+{% endraw %}
+    # OpenID Connect configuration for EGI Check-in AAI
+    OIDCResponseType "code"
+    OIDCClaimPrefix "OIDC-"
+    OIDCClaimDelimiter ;
+    OIDCScope "openid"
+    OIDCProviderMetadataURL https://aai-dev.egi.eu/oidc/.well-known/openid-configuration
+    OIDCClientID {{ secrets_egi_client_id }}
```
Comment (dougszumski) [inline etc/kayobe/kolla/config/keystone/wsgi-keystone.conf:28]: nit: Is this one actually a secret?

**Review by dougszumski — APPROVED** [review]

File: etc/kayobe/kolla/config/keystone/wsgi-keystone.conf (hunk 2 above)
Comment (yankcrime) [reply etc/kayobe/kolla/config/keystone/wsgi-keystone.conf:28]: Not necessarily secret, but arguably sensitive.

File: etc/kayobe/kolla/config/keystone/wsgi-keystone.conf (hunk 1 above)
Comment (yankcrime) [reply etc/kayobe/kolla/config/keystone/wsgi-keystone.conf:7]: Inline you mean, or as part of the commit message?

File: etc/kayobe/kolla/config/horizon/local_settings [hunk 3]
```diff
@@ -0,0 +1,824 @@
+{% raw %}
```
Comment (markgoddard) [inline etc/kayobe/kolla/config/horizon/local_settings:1]:
```
$ git diff ../kolla-ansible/ansible/roles/horizon/templates/local_settings.j2 etc/kayobe/kolla/config/horizon/local_settings 
diff --git a/../kolla-ansible/ansible/roles/horizon/templates/local_settings.j2 b/etc/kayobe/kolla/config/horizon/local_settings
index a7926ad..64993fe 100644
--- a/../kolla-ansible/ansible/roles/horizon/templates/local_settings.j2
+++ b/etc/kayobe/kolla/config/horizon/local_settings
@@ -1,3 +1,4 @@
+{% raw %}
 # -*- coding: utf-8 -*-
 
 import os
@@ -190,7 +191,7 @@ AVAILABLE_REGIONS = [
 
 OPENSTACK_HOST = "{{ kolla_internal_fqdn }}"
 
-OPENSTACK_KEYSTONE_URL = "{{ keystone_internal_url }}"
+OPENSTACK_KEYSTONE_URL = "{{ keystone_public_url }}"
 OPENSTACK_KEYSTONE_DEFAULT_ROLE = "{{ keystone_default_user_role }}"
 
 # Enables keystone web single-sign-on if set to True.
@@ -814,3 +815,10 @@ REST_API_REQUIRED_SETTINGS = ['OPENSTACK_HYPERVISOR_FEATURES',
 # before loading any data into the admin views, set the following attribute to
 # True
 #ADMIN_FILTER_DATA_FIRST=False
+{% endraw %}
+
+WEBSSO_ENABLED = True
+WEBSSO_CHOICES = (
+    ("credentials", _("Keystone Credentials")),
+    ("oidc", _("EGI AAI Check-in")),
+)
```

File: etc/kayobe/kolla/config/keystone/wsgi-keystone.conf [hunk 4]
```diff
@@ -0,0 +1,63 @@
+{% raw %}
```
Comment (markgoddard) [inline etc/kayobe/kolla/config/keystone/wsgi-keystone.conf:1]:
```
git diff ../kolla-ansible/ansible/roles/keystone/templates/wsgi-keystone.conf.j2 etc/kayobe/kolla/config/keystone/wsgi-keystone.conf
diff --git a/../kolla-ansible/ansible/roles/keystone/templates/wsgi-keystone.conf.j2 b/etc/kayobe/kolla/config/keystone/wsgi-keystone.conf
index 83b297a..0f23e98 100644
--- a/../kolla-ansible/ansible/roles/keystone/templates/wsgi-keystone.conf.j2
+++ b/etc/kayobe/kolla/config/keystone/wsgi-keystone.conf
@@ -1,3 +1,4 @@
+{% raw %}
 {% set keystone_log_dir = '/var/log/kolla/keystone' %}
 {% set python_path = '/usr/lib/python2.7/site-packages' if kolla_install_type == 'binary' else '/var/lib/kolla/venv/lib/python2.7/site-packages' %}
 Listen {{ api_interface_address }}:{{ keystone_public_port }}
@@ -17,8 +18,35 @@ TraceEnable off
     ErrorLog "{{ keystone_log_dir }}/keystone-apache-public-error.log"
     LogFormat "%{X-Forwarded-For}i %l %u %t \"%r\" %>s %b %D \"%{Referer}i\" \"%{User-Agent}i\"" logformat
     CustomLog "{{ keystone_log_dir }}/keystone-apache-public-access.log" logformat
+{% endraw %}
+    # OpenID Connect configuration for EGI Check-in AAI
+    OIDCResponseType "code"
+    OIDCClaimPrefix "OIDC-"
+    OIDCClaimDelimiter ;
+    OIDCScope "openid"
+    OIDCProviderMetadataURL https://aai-dev.egi.eu/oidc/.well-known/openid-configuration
+    OIDCClientID {{ secrets_egi_client_id }}
+    OIDCClientSecret {{ secrets_egi_client_secret }}
+    OIDCCryptoPassphrase {{ secrets_egi_crypto_passphrase }}
+    OIDCRedirectURI http://{% raw %}{{ kolla_external_fqdn }}:{{ keystone_public_port }}{% endraw %}/v3/auth/OS-FEDERATION/websso/oidc/redirect
+
+    # OAuth for CLI access
+    OIDCOAuthIntrospectionEndpoint  https://aai-dev.egi.eu/oidc/introspect
+    OIDCOAuthClientID {{ secrets_egi_client_id }}
+    OIDCOAuthClientSecret {{ secrets_egi_client_secret }}
+
+    <Location ~ "/v3/auth/OS-FEDERATION/websso/oidc">
+            AuthType  openid-connect
+            Require   valid-user
+    </Location>
+
+    <Location ~ "/v3/OS-FEDERATION/identity_providers/egi.eu/protocols/oidc/auth">
+            Authtype oauth20
+            Require   valid-user
+    </Location>
 </VirtualHost>
 
+{% raw %}
 <VirtualHost *:{{ keystone_admin_port }}>
     WSGIDaemonProcess keystone-admin processes={{ openstack_service_workers }} threads=1 user=keystone group=keystone display-name=%{GROUP} python-path={{ python_path }}
     WSGIProcessGroup keystone-admin
@@ -32,3 +60,4 @@ TraceEnable off
     LogFormat "%{X-Forwarded-For}i %l %u %t \"%r\" %>s %b %D \"%{Referer}i\" \"%{User-Agent}i\"" logformat
     CustomLog "{{ keystone_log_dir }}/keystone-apache-admin-access.log" logformat
 </VirtualHost>
+{% endraw %}
```

File: etc/kayobe/kolla/config/keystone/wsgi-keystone.conf (hunk 1 above)
Comment (dougszumski) [reply etc/kayobe/kolla/config/keystone/wsgi-keystone.conf:7]: I was just thinking an inline comment marking your customisation from the base config, or perhaps two commits, one of the base config file, and the other with the customisation for ```git blame```. Not to worry too much.

Comment (yankcrime) [reply etc/kayobe/kolla/config/keystone/wsgi-keystone.conf:7]: I thought about two discrete commit but then there's still no way to guarantee provenance.  Point taken about traceability with `git blame` though, at least in order to narrow scope.

**Review by markgoddard — APPROVED** [review]
The diffs against kolla-ansible templates look sane, although I don't have enough knowledge of the technologies used here to be sure.

**Pull request merged**